### ISSUE TITLE
Segmentation fault with rlm_sql_iodbc and MSSQL

### ISSUE THREAD

**Issue by tomplast** [FreeRADIUS/freeradius-server#1395]
I'm using rlm_sql_iodbc module with MSSQL and repeatedly I can make FreeRadius segfault when sending Access-Request packages. I get the following message whenever this happens: 

rlm_sql (sql): Closing connection (2): Hit idle_timeout, was idle for 119 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (6): Hit idle_timeout, was idle for 119 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (3): Hit idle_timeout, was idle for 119 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (4): Hit idle_timeout, was idle for 116 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (7): Hit idle_timeout, was idle for 115 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (0): Hit idle_timeout, was idle for 115 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (8): Hit idle_timeout, was idle for 115 seconds
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (5): Hit idle_timeout, was idle for 103 seconds
rlm_sql (sql): You probably need to lower "min"
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (9): Hit idle_timeout, was idle for 103 seconds
rlm_sql (sql): You probably need to lower "min"
rlm_sql_iodbc: Socket destructor called, closing socket
rlm_sql (sql): Closing connection (1): Hit idle_timeout, was idle for 103 seconds
rlm_sql (sql): You probably need to lower "min"
rlm_sql_iodbc: Socket destructor called, closing socket

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff172783b in SQLFreeStmt ()
   from /usr/lib/x86_64-linux-gnu/libiodbc.so.2

I'm using Debian 8.2 x86_64 and FreeRadius 3.0.10 (ftp://ftp.freeradius.org/pub/freeradius/freeradius-server-3.0.10.tar.gz). Any ideas why this is happening and how this can be resolved?

**Comment by arr2036** [comment]
Follow instructions here for reporting bugs:

	https://github.com/FreeRADIUS/freeradius-server/blob/v3.1.x/doc/bugs

**Comment by arr2036** [comment]
Usually just running ``gdb --args <path to radius> -f`` and running ``bt`` which is crashes provides us with enough info.

**Comment by arr2036** [comment]
Closing this issue as no feedback from @tomplast.  Will re-open if backtrace is received.

**Comment by JanderII** [comment]
We are seeing the same issue with accounting packets. We are using Ubuntu 16.04 and the FreeRADIUS 3.0.12-ppa1 package from http://ppa.launchpad.net/freeradius/stable-3.0/ubuntu.

We have a combination of segfaults and sigaborts apparently due to memory corruption. This corruption affects the closing of database connections, exactly as @tomplast had seen. The program crashes both randomly, and during a service shutdown/restart. Thankfully it is being run under systemd so I have been able to capture core dumps.

Here's an example from the RADIUS log:

`Sat Apr 29 05:40:01 2017 : Info: rlm_sql (sql): Opening additional connection (108), 7 of 7 pending slots used
Sat Apr 29 05:40:31 2017 : Info: rlm_sql (sql): Closing connection (65), from 22 unused connections
Sat Apr 29 05:40:46 2017 : Info: rlm_sql (sql): Closing connection (38), from 21 unused connections
Sat Apr 29 05:40:53 2017 : Info: rlm_sql (sql): Closing connection (101), from 19 unused connections
Sat Apr 29 05:40:56 2017 : Info: rlm_sql (sql): Closing connection (91), from 19 unused connections
Sat Apr 29 05:40:57 2017 : Info: rlm_sql (sql): Closing connection (95), from 18 unused connections
Sat Apr 29 05:40:58 2017 : Info: rlm_sql (sql): Closing connection (105), from 17 unused connections
Sat Apr 29 05:40:59 2017 : Info: rlm_sql (sql): Closing connection (85), from 15 unused connections
Sat Apr 29 05:41:00 2017 : Info: rlm_sql (sql): Closing connection (106), from 15 unused connections
Sat Apr 29 05:41:01 2017 : Info: rlm_sql (sql): Closing connection (100), from 1 unused connections
Sat Apr 29 05:41:07 2017 : Error: Cannot update core dump limit: Operation not permitted
Sat Apr 29 05:41:07 2017 : Info: Core dumps are enabled`

I've attached a "gdb bt full" for reference. 

[radius-gdb.txt](https://github.com/FreeRADIUS/freeradius-server/files/987117/radius-gdb.txt)

If we look at frame # 6 we can see what's going on:

`#6  0x00007fb3f3c98e99 in _iodbcdm_popsqlerr (herr=<optimised out>)
    at herr.c:116
        list = <optimised out>
        next = 0x2d3a007364726f77`

The next pointer is way out of memory, and is actually ASCII text: "-:\0sdrow". This would appear to be part of the SQL insert statement we have for our accounting, specifically part of:

"@p25=%{%{Acct-Input-Gigawords}:-0}, \
@p26=%{%{Acct-Output-Gigawords}:-0}"

If I look at frame # 8:

`#8  0x00007fb3f3c9d97b in SQLFreeStmt (hstmt=0x7fb3cc018950,
    fOption=fOption@entry=1) at hstmt.c:640
        pstmt = 0x7fb3cc018950
        retcode = 0`

and then look at the memory for hstmt:

`(gdb) x/100x hstmt
0x7fb3cc018950: 0x00000003      0x00000000      0xcc06d1cc      0x00007fb3
0x7fb3cc018960: 0x00000015      0x00000000      0x00000000      0x00000000
0x7fb3cc018970: 0x00000002      0x00000001      0x0257e090      0x00000000
0x7fb3cc018980: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc018990: 0x00000000      0x00000001      0x00000000      0x00000000
0x7fb3cc0189a0: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc0189b0: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc0189c0: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc0189d0: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc0189e0: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc0189f0: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc018a00: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc018a10: 0x00000000      0x80000000      0x00000080      0x00000000
0x7fb3cc018a20: 0x00000000      0x00000000      0x000000b1      0x00000000
0x7fb3cc018a30: 0xcc001790      0x00007fb3      0xcc02aaf0      0x00007fb3
0x7fb3cc018a40: 0x00000000      0x00000000      0xcc001270      0x00007fb3
0x7fb3cc018a50: 0x00000000      0x00000000      0x00000000      0x00000000
0x7fb3cc018a60: 0x00000000      0x00000000      0xf78e1f11      0x00007fb3
0x7fb3cc018a70: 0x0000002d      0x00000000      0x00000000      0x00000000
0x7fb3cc018a80: 0x00000000      0x00000000      0xcc000078      0x00007fb3
0x7fb3cc018a90: 0x65746e49      0x2d6d6972      0x61647055      0x00006574
0x7fb3cc018aa0: 0xcc0129e0      0x00007fb3      0xcc000078      0x00007fb3
0x7fb3cc018ab0: 0x000000a0      0x00000000      0x00000024      0x00000000
0x7fb3cc018ac0: 0xcc06d0d0      0x00007fb3      0xcc000088      0x00007fb3
0x7fb3cc018ad0: 0x00000950      0x00000000      0x000000c4      0x00000000`

we can see odd data in memory 0x7fb3cc018a90. This again is ASCII text, which decodes to "Interim-Update".

So what we appear to have is some form of memory corruption whereby SQL or RADIUS data is stomping over elements within the IODBC connections. When these connections are closed down they are accessing these invalid pointers, resulting in segfaults or sigaborts. 

When I run the server in single threaded mode "-t" it does not suffer from the random segfaulting, however it still crashes on shutdown/restart with similar corruption to the iodbc data.

At this time I can only conclude that there's a use-after-free style problem going on.
